lib/grpc/client/adapters/mint/connection_process/connection_process.ex

```diff
@@ -0,0 +1,395 @@
+    {:reply, {:error, "the connection is closed"}, state}
+  end
+
+  def handle_call(
+        {:request, method, path, headers, :stream, opts},
+        _from,
+        state
+      ) do
+    case Mint.HTTP.request(state.conn, method, path, headers, :stream) do
+      {:ok, conn, request_ref} ->
+        new_state =
+          state
+          |> State.update_conn(conn)
+          |> State.put_empty_ref_state(request_ref, opts[:stream_response_pid])
+
+        {:reply, {:ok, %{request_ref: request_ref}}, new_state}
+
+      {:error, conn, reason} ->
+        new_state = State.update_conn(state, conn)
+        {:reply, {:error, reason}, new_state}
+    end
+  end
+
+  def handle_call(
+        {:request, method, path, headers, body, opts},
+        _from,
+        state
+      ) do
+    case Mint.HTTP.request(state.conn, method, path, headers, :stream) do
+      {:ok, conn, request_ref} ->
+        queue = :queue.in({request_ref, body, nil}, state.request_stream_queue)
+
+        new_state =
+          state
+          |> State.update_conn(conn)
+          |> State.update_request_stream_queue(queue)
+          |> State.put_empty_ref_state(request_ref, opts[:stream_response_pid])
+
+        {:reply, {:ok, %{request_ref: request_ref}}, new_state,
+         {:continue, :process_request_stream_queue}}
+
+      {:error, conn, reason} ->
+        new_state = State.update_conn(state, conn)
+        {:reply, {:error, reason}, new_state}
+    end
+  end
+
+  def handle_call({:stream_body, request_ref, :eof}, _from, state) do
+    case Mint.HTTP.stream_request_body(state.conn, request_ref, :eof) do
+      {:ok, conn} ->
+        {:reply, :ok, State.update_conn(state, conn)}
+
+      {:error, conn, error} ->
+        {:reply, {:error, error}, State.update_conn(state, conn)}
+    end
+  end
+
+  def handle_call({:stream_body, request_ref, body}, from, state) do
+    queue = :queue.in({request_ref, body, from}, state.request_stream_queue)
+
+    {:noreply, State.update_request_stream_queue(state, queue),
+     {:continue, :process_request_stream_queue}}
+  end
+
+  def handle_call({:cancel_request, request_ref}, _from, state) do
+    state = process_response({:done, request_ref}, state)
+
+    case Mint.HTTP2.cancel_request(state.conn, request_ref) do
+      {:ok, conn} -> {:reply, :ok, State.update_conn(state, conn)}
+      {:error, conn, error} -> {:reply, {:error, error}, State.update_conn(state, conn)}
+    end
+  end
+
+  @impl true
+  def handle_info(message, state) do
+    case Mint.HTTP.stream(state.conn, message) do
+      :unknown ->
+        Logger.debug(fn -> "Received unknown message: " <> inspect(message) end)
+        {:noreply, state}
+
+      {:ok, conn, responses} ->
+        state = State.update_conn(state, conn)
+
+        state =
+          if state.requests == %{},
+            do: state,
+            else: Enum.reduce(responses, state, &process_response/2)
```

Comment:
Just to avoid map-equality, which can be a bit misleading:

```suggestion
          case state.requests do
            requests when map_size(requests) == 0 ->
              state
            _ ->
              Enum.reduce(responses, state, &process_response/2)
          end
```

Also, I wonder if this isn't a race condition for a response arriving before the corresponding request is saved.
Just to confirm it isn't, a request is sent by this same process, right? If so, this isn't a race condition because the process will process messages sequentially anyway.

Reply:
Gotcha! (for the map - I just read about that on elixir forum - TIL) 
For the race condition: yeah. That's not a problem we send the request by this same process indeed.